src/Microsoft.AspNet.OData.Shared/Query/ExpandQueryBuilder.cs

```diff
@@ -0,0 +1,121 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ExpandQueryBuilder.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved. 
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Formatter;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Exposes the ability to generate a $expand query string from a payload object.
+    /// </summary>
+    public class ExpandQueryBuilder : IExpandQueryBuilder
+    {
+        /// <inheritdoc />
+        public virtual string GenerateExpandQueryString(object value, IEdmModel model)
+        {
+            if (value == null)
+            {
+                throw Error.ArgumentNull(nameof(value));
+            }
+
+            if (model == null)
+            {
+                throw Error.ArgumentNull(nameof(model));
+            }
+
+            return GenerateExpandQueryStringInternal(value, model, true);
+        }
+
+        private string GenerateExpandQueryStringInternal(object value, IEdmModel model, bool isFirstExpand)
```

Comment:
Given you explanation about `isFirstExpand`, you could rename this to `isNestedExpand` and pass a value of `false` from `GenerateExpandQueryString` method. For most people, nested in the context of `$expand` query option is easier to relate with. `IsFirstExpand` on the other hand needs an explanation.